api_core/google/api_core/grpc_helpers.py

```diff
@@ -64,6 +64,18 @@ def error_remapped_callable(*args, **kwargs):
 class _StreamingResponseIterator(grpc.Call):
     def __init__(self, wrapped):
         self._wrapped = wrapped
+        self._stored_first_result = False
+
+        # This iterator is used in a retry context, and returned outside after init.
+        # gRPC will not throw an exception until the stream is consumed, so we need 
+        # to retrieve the first result, in order to fail, in order to trigger a retry.
+        try:
+            self._first_result = six.next(self._wrapped)
+            self._stored_first_result = True
+        except StopIteration:
```

Comment:
Will this work for the WatchStream? For Watch, the client has to send the first request. We will not receive a response until after this request has been processed.

Reply:
Watch uses `ResumableBidiRpc` and has its own way of managing recovery. This will have an effect on things like `query` though.